### PR TITLE
Further refinements to CSS/Style gradient code

### PR DESCRIPTION
**Pull request by weinig**
#### 9008eb572f870b3b669cd5995310538f88642bc1
<pre>
Further refinements to CSS/Style gradient code
<a href="https://bugs.webkit.org/show_bug.cgi?id=281351">https://bugs.webkit.org/show_bug.cgi?id=281351</a>

Reviewed by Darin Adler.

- Adds structural wrapper representing CSS function syntax
  (like `linear-gradient()`) and adopts them in the gradient
  code, allowing automatic serialization of the syntax.

- Adds Space/CommaSeparatedArray type to represent fixed
  size repeated elements of the same type. Adopted where
  appropriate in gradient code.

- Replaced the 7 Gradient CSSValue subclasses with a single
  CSSGradientValue subclass that uses a std::variant to
  dispatch to the appropriate value.

- Moved Style::Gradient -&gt; WebCore::Gradient conversion code
  to StyleGradient.h/cpp (from StyleGradientImage.h/cpp) as
  it is purely a functional transformation and not specific
  to using the gradient as an image.

* Source/WTF/wtf/StdLibExtras.h:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSGradientValue.h:
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/values/CSSGradient.cpp:
* Source/WebCore/css/values/CSSGradient.h:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/html/HTMLInputElement.cpp:
* Source/WebCore/rendering/style/StyleGradientImage.cpp:
* Source/WebCore/rendering/style/StyleGradientImage.h:
* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StyleScopeRuleSets.cpp:
* Source/WebCore/style/StyleScopeRuleSets.h:
* Source/WebCore/style/values/StyleGradient.cpp:
* Source/WebCore/style/values/StyleGradient.h:
* Source/WebCore/style/values/StylePosition.h:
* Source/WebCore/style/values/StyleValueTypes.h:

Canonical link: <a href="https://commits.webkit.org/285075@main">https://commits.webkit.org/285075@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/f95b6e3701a72090b94b63a06cfdbd8986c6b196

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71423 "14 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50836 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24196 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75534 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22628 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73538 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58636 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22447 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56420 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14886 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74489 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46147 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61530 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36860 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42810 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18995 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20969 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/64548 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64709 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19361 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77253 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/70672 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15657 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18534 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64132 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15700 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61570 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64124 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15802 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12275 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5907 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/92458 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46636 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1415 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20387 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47707 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48990 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47449 "Built successfully") | | |

### CODE REVIEW

**Comment by weinig** [comment]
I had to use FunctionNotation<>, rather than just Function<> due to ambiguity with WTF::Function<> which gets injected into the global namespace.

File: Source/WTF/wtf/StdLibExtras.h [hunk 1]
```diff
@@ -432,6 +432,11 @@ bool findBitInWord(T word, size_t& startOrResultIndex, size_t endIndex, bool val
     return false;
 }
 
+// Used to check if a variadic list of compile time predicates are all true.
```
Comment (darinadler) [inline Source/WTF/wtf/StdLibExtras.h:435]: I find this implementation unclear but the use is clear. Is there a coming standard for this or is this just a useful thing they are not planing to add?

File: Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp [hunk 2]
```diff
@@ -597,7 +603,7 @@ static RefPtr<CSSValue> consumePrefixedRadialGradient(CSSParserTokenRange& range
 // MARK: <linear-gradient()> | <repeating-linear-gradient()>
 // https://drafts.csswg.org/css-images-4/#linear-gradients
 
-static RefPtr<CSSValue> consumeLinearGradient(CSSParserTokenRange& range, const CSSParserContext& context, CSS::GradientRepeat repeating)
+template<CSSValueID Name> static RefPtr<CSSValue> consumeLinearGradient(CSSParserTokenRange& range, const CSSParserContext& context)
```
Comment (darinadler) [inline Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:606]: How did you decide on template argument vs. function argument?

File: Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp [hunk 3]
```diff
@@ -693,20 +699,21 @@ static RefPtr<CSSValue> consumeLinearGradient(CSSParserTokenRange& range, const
...
+            }
         }
     );
 }
 
 // MARK: <radial-gradient()> | <repeating-radial-gradient()>
 // https://drafts.csswg.org/css-images-4/#radial-gradients
 
-static RefPtr<CSSValue> consumeRadialGradient(CSSParserTokenRange& range, const CSSParserContext& context, CSS::GradientRepeat repeating)
+template<CSSValueID Name> static RefPtr<CSSValue> consumeRadialGradient(CSSParserTokenRange& range, const CSSParserContext& context)
```
Comment (darinadler) [inline Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:716]: How did you decide on template argument vs. function argument?

**Review by darinadler — APPROVED** [review]

File: Source/WTF/wtf/StdLibExtras.h (hunk 1 above)
Comment (weinig) [reply Source/WTF/wtf/StdLibExtras.h:435]: I am not sure if there is an upcoming standard, but it is used commonly in the standard library so should be. I’ll check it out and see if it has been proposed.

File: Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp (hunk 2 above)
Comment (weinig) [reply Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:606]: I need the template argument to be a compile time constant for use in the return value, so I chose template.

File: Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp (hunk 3 above)
Comment (weinig) [reply Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:716]: Same as above. Needed Name to be a template argument in FunctionNotation.

**Pull request merged**